markdown.dtx

```diff
@@ -24070,873 +24083,2822 @@ parsers.spnl                   = parsers.optionalspace
-              if is_shorter then
-                  min_prefix_length, min_prefix = prefix_length, prefix
-              end
-              return line
-          end)
-          str = str:gsub('^' .. min_prefix, '')
-      end
-%    \end{macrocode}
-% \par
+  local function render_link_or_image(t, opening_index, closing_index, content_end_index, reference)
+    process_emphasis(t, opening_index, content_end_index)
+    local mapped = collect_emphasis_content(t, opening_index + 1, content_end_index - 1)
+
+    local rendered = {}
+    if (t[opening_index].element == "link") then
+      rendered = writer.link(mapped, reference.url, reference.title, reference.attributes)
+    end
+
+    if (t[opening_index].element == "image") then
+      rendered = writer.image(mapped, reference.url, reference.title, reference.attributes)
+    end
+
+    t[opening_index].rendered = rendered
+    delete_parsed_content_in_range(t, opening_index + 1, closing_index)
+    empty_content_in_range(t, opening_index, closing_index)
+    disable_previous_link_openers(t, opening_index)
+    disable_range(t, opening_index, closing_index)
+  end
+
+%  \end{macrocode}
 % \begin{markdown}
 %
-% If the parser is top-level and the \Opt{texComments} or \Opt{hybrid} Lua
-% options are enabled, we will strip all plain \TeX{} comments from the input
-% string `str` together with the trailing newline characters.
+% Match the link destination of an inline link at index `closing_index` in table `t`
+% when `match_reference` is true. Additionally, match attributes when the option
+% \Opt{linkAttributes} is enabled.
 %
 % \end{markdown}
 %  \begin{macrocode}
-      if toplevel and (options.texComments or options.hybrid) then
-        str = lpeg.match(Ct(parsers.commented_line^1), str)
-        str = util.rope_to_string(str)
-      end
-      local res = lpeg.match(grammar(), str)
-      if res == nil then
-        error(format("%s failed on:\n%s", name, str:sub(1,20)))
-      else
-        return res
-      end
+  local function resolve_inline_following_content(t, closing_index, match_reference, match_link_attributes)
+    local content = ""
+    for i = closing_index + 1, #t do
+      content = content .. t[i].content
     end
-  end
 
-  self.create_parser("parse_blocks",
-                     function()
-                       return parsers.blocks
-                     end, true)
+    local matching_content = parsers.succeed
 
-  self.create_parser("parse_blocks_nested",
-                     function()
-                       return parsers.blocks_nested
-                     end, false)
+    if match_reference then
+      matching_content = matching_content * parsers.inline_direct_ref_inside
+    end
 
-  self.create_parser("parse_inlines",
-                     function()
-                       return parsers.inlines
-                     end, false)
+    if match_link_attributes then
+      matching_content = matching_content * Cg(Ct(parsers.attributes^-1), "attributes")
+    end
```

Comment:
@lostenderman: The purpose of built-in syntax extensions was to separate the syntax extensions from the base reader in order to make the code more modular and easier to reason about. In essence, the option `options.linkAttributes` should not be accessed from the base markdown reader. All code for `options.linkAttributes` should be encapsulated in `M.extensions.link_attributes`. Not sure how to get there but the current implementation is a step down in terms of code quality.